subsys/mgmt/hawkbit/hawkbit.c

```diff
@@ -986,235 +891,454 @@ void hawkbit_reboot(void)
 	sys_reboot(SYS_REBOOT_WARM);
 }
 
-enum hawkbit_response hawkbit_probe(void)
+static void s_start(void *o)
 {
-	int ret;
-	int32_t action_id;
-	int32_t file_size = 0;
-	struct flash_img_check fic;
-	char device_id[DEVICE_ID_HEX_MAX_SIZE] = { 0 },
-	     cancel_base[CANCEL_BASE_SIZE] = { 0 },
-	     download_http[DOWNLOAD_HTTP_SIZE] = { 0 },
-	     deployment_base[DEPLOYMENT_BASE_SIZE] = { 0 },
-	     firmware_version[BOOT_IMG_VER_STRLEN_MAX] = { 0 };
+	struct s_object *s = (struct s_object *)o;
 
 	if (k_sem_take(&probe_sem, K_NO_WAIT) != 0) {
-		return HAWKBIT_PROBE_IN_PROGRESS;
+		LOG_INF("hawkBit is already running");
+		smf_set_terminate(SMF_CTX(s), HAWKBIT_PROBE_IN_PROGRESS);
+		return;
 	}
 
-	memset(&hb_context, 0, sizeof(hb_context));
-	hb_context.response_data = malloc(RESPONSE_BUFFER_SIZE);
-
 	if (!boot_is_img_confirmed()) {
 		LOG_ERR("Current image is not confirmed");
-		hb_context.code_status = HAWKBIT_UNCONFIRMED_IMAGE;
-		goto error;
+		k_sem_give(&probe_sem);
+		smf_set_terminate(SMF_CTX(s), HAWKBIT_UNCONFIRMED_IMAGE);
+		return;
 	}
 
-	if (!hawkbit_get_firmware_version(firmware_version, BOOT_IMG_VER_STRLEN_MAX)) {
-		hb_context.code_status = HAWKBIT_METADATA_ERROR;
-		goto error;
+	if (!hawkbit_get_device_identity(s->device_id, DEVICE_ID_HEX_MAX_SIZE)) {
+		k_sem_give(&probe_sem);
+		smf_set_terminate(SMF_CTX(s), HAWKBIT_METADATA_ERROR);
+		return;
 	}
+}
+
+static void s_end(void *o)
+{
+	struct s_object *s = (struct s_object *)o;
+
+	k_sem_give(&probe_sem);
+	smf_set_terminate(SMF_CTX(s), s->hb_context.code_status);
+}
+
+static void s_http_start(void *o)
+{
+	struct s_object *s = (struct s_object *)o;
 
-	if (!hawkbit_get_device_identity(device_id, DEVICE_ID_HEX_MAX_SIZE)) {
-		hb_context.code_status = HAWKBIT_METADATA_ERROR;
-		goto error;
+	if (!start_http_client(&s->hb_context.sock)) {
+		s->hb_context.code_status = HAWKBIT_NETWORKING_ERROR;
+		smf_set_state(SMF_CTX(s), NULL);
```

Comment:
I have a dislike of this construct as it doesn't follow UML rules for state machines, and I'm hoping to put in an RFC for making SMF UML-compliant.

Would it be possible to make an error-handling state outside the other states and at the same level as the root state, that you transition to on an error? This would cause all exit actions to be called, plus the entry action of the error-handler state, which can set_smf_terminate() in its entry action.

Reply:
I think so, I'm looking on my phone so not able to do a full examination.

Is s_terminate the entry action or the run action? If it's the run action you'd need to post an event to get it to run, I think?

Reply:
it's an entry action.

Reply:
Sounds good. Thank you.